File: src/lib/attention.ts
// Pure triage helpers for the sidebar, the details pane and the attention
// strip: status vocabulary → label/colour, context-pressure levels, the
// "needs attention" predicate, per-row severity for sorting projects by their
// worst child, stuck-transition detection between two store snapshots, and
// the small display formatters for the outcome fields (elapsed, last prompt).
//
// Everything here is side-effect free so it is unit-testable without
// mounting a component, and so Sidebar.svelte can run each helper ONCE per
// store change inside a $derived rather than per row.
import {
  CLAUDE_STATUSES,
  STUCK_KINDS,
  type CiStatus,
  type ClaudeStatus,
  type SessionRow,
  type StuckKind,
} from './sessions';

// ── status vocabulary ──

export function isClaudeStatus(v: unknown): v is ClaudeStatus {
  return typeof v === 'string' && (CLAUDE_STATUSES as readonly string[]).includes(v);
}

export function isStuckKind(v: unknown): v is StuckKind {
  return typeof v === 'string' && (STUCK_KINDS as readonly string[]).includes(v);
}

const STATUS_COLOR: Record<ClaudeStatus, string> = {
  working: '#50c86e', // green — active
  blocked: '#f0b429', // yellow — needs input
  completed: '#6c8ebf', // blue — done
  failed: '#e64a4a', // red
  stopped: '#888', // grey — stopped by hook or user
  idle: '#888', // grey
};

const STATUS_LABEL: Record<ClaudeStatus, string> = {
  working: '⚡ working',
  blocked: '⏸ blocked',
  completed: '✓ done',
  failed: '✗ failed',
  stopped: '■ stopped',
  idle: '· idle',
};

export function claudeStatusColor(status: ClaudeStatus | null): string {
  return status && isClaudeStatus(status) ? STATUS_COLOR[status] : 'transparent';
}

export function claudeStatusLabel(status: ClaudeStatus | null): string {
  return status && isClaudeStatus(status) ? STATUS_LABEL[status] : '';
}

const STUCK_LABEL: Record<StuckKind, string> = {
  auth_menu: 'auth menu',
  reconnect: 'reconnecting',
  trust_prompt: 'trust prompt',
  oom: 'out of memory',
  press_enter: 'press Enter',
};

/** Human label for a stuck kind, e.g. `press_enter` → "press Enter". */
export function stuckKindLabel(kind: StuckKind | null): string {
  return kind && isStuckKind(kind) ? STUCK_LABEL[kind] : '';
}

/** Red, always — the stuck chip outranks whatever claude_status says. */
export const STUCK_COLOR = '#e64a4a';

// ── context pressure ──

export type ContextLevel = 'ok' | 'warn' | 'crit';

export const CONTEXT_WARN_PCT = 70;
export const CONTEXT_CRIT_PCT = 90;

export function contextLevel(pct: number | null): ContextLevel | null {
  if (pct === null || !Number.isFinite(pct)) return null;
  if (pct >= CONTEXT_CRIT_PCT) return 'crit';
  if (pct >= CONTEXT_WARN_PCT) return 'warn';
  return 'ok';
}

export function contextColor(level: ContextLevel | null): string {
  switch (level) {
    case 'crit':
      return '#e64a4a';
    case 'warn':
      return '#d29b4a';
    case 'ok':
      return '#50c86e';
    default:
      return 'transparent';
  }
}

// ── triage ranking (P13) ──

export interface AttentionOptions {
  /** Work sessions idle at least this long (seconds) need a nudge. 0 = off. */
  idleSecs: number;
  /** Unix seconds "now" (injected so tests are deterministic). */
  now: number;
}

export const DEFAULT_ATTENTION_IDLE_MINUTES = 30;

/** Triage buckets, most urgent first. `classify()` puts a row in exactly one,
 *  and everything that orders sessions reads this one list — the sidebar's
 *  "Needs you" queue, the project sort below, later the quick switcher and the
 *  digest — so those orderings cannot drift apart.
 *
 *  Two buckets are reachable but stay empty until Wave 1 A2 lands its columns:
 *   - `waiting` is driven by `claude_status === 'blocked'` alone. A2's
 *     `waiting_for` will separate a permission prompt from a question and let
 *     the age weighting apply per kind.
 *   - `done_unread` needs `last_viewed_at` and its `touch_session_viewed`
 *     writer, so nothing matches it today. */
export const TRIAGE_BUCKETS = [
  'waiting',
  'stuck',
  'failed',
  'done_unread',
  'lifecycle',
  'idle_long',
  'working',
  'idle',
] as const;

export type TriageBucket = (typeof TRIAGE_BUCKETS)[number];

/** Buckets the "Needs you" FILTER shows. `idle_long` is in: the toggle is the
 *  only surface for the operator-configured idle nudge, so leaving it out
 *  would delete that reach and reduce `attentionIdleMinutes` to a sort knob.
 *  `working` and `idle` are never in it. */
export const NEEDS_YOU_BUCKETS: readonly TriageBucket[] = TRIAGE_BUCKETS.slice(0, 6);

/** Buckets the "Needs you" COUNTER reports — deliberately one narrower than
 *  the filter, excluding `idle_long`.
 *
 *  The divergence is intentional, not an oversight. The pill answers "which
 *  sessions need me NOW"; on a fleet of ~60 sessions most are idle, so
 *  counting them would read "Needs you (34)" and the number would stop
 *  meaning anything. The rows are still one toggle away, because the filter
 *  above does include them. Do not "reconcile" these two sets. */
export const NEEDS_YOU_COUNTED_BUCKETS: readonly TriageBucket[] = TRIAGE_BUCKETS.slice(0, 5);

const NEEDS_YOU = new Set<TriageBucket>(NEEDS_YOU_BUCKETS);
const NEEDS_YOU_COUNTED = new Set<TriageBucket>(NEEDS_YOU_COUNTED_BUCKETS);

/** Age is capped so that no wait, however long, lets a row jump its bucket. */
const AGE_CAP_SECS = 1_000_000;

export interface TriageRank {
  bucket: TriageBucket;
  /** Index into TRIAGE_BUCKETS; 0 is the most urgent. */
  order: number;
  /** Seconds spent in this state; 0 when the row carries no usable stamp. */
  ageSecs: number;
  /** Sort weight, higher = more urgent. The bucket dominates, age breaks ties. */
  score: number;
}

/** A2: `waiting_for` will distinguish permission, question and elicitation.
 *  Until then a blocked session is the only thing known to await the user. */
function isWaiting(s: SessionRow): boolean {
  return s.claude_status === 'blocked';
}

/** A2: needs `last_viewed_at`, so nothing is done-unread yet. */
function isDoneUnread(_s: SessionRow): boolean {
  return false;
}

function isLifecycleBroken(s: SessionRow): boolean {
  if (s.safe_kill_state === 'failed' || s.safe_kill_state === 'requested') return true;
  return s.status === 'ghost' || s.lost_at !== null;
}

function isIdleLong(s: SessionRow, opts: AttentionOptions): boolean {
  if (opts.idleSecs <= 0) return false;
  if (s.kind !== 'work' && s.kind !== 'review') return false;
  if (s.idle_since === null) return false;
  return opts.now - s.idle_since >= opts.idleSecs;
}

/** The single classifier: the order of these checks IS the bucket order. */
export function classify(s: SessionRow, opts: AttentionOptions): TriageBucket {
  if (isWaiting(s)) return 'waiting';
  if (s.stuck_kind) return 'stuck';
  if (s.claude_status === 'failed') return 'failed';
  if (isDoneUnread(s)) return 'done_unread';
  if (isLifecycleBroken(s)) return 'lifecycle';
  if (isIdleLong(s, opts)) return 'idle_long';
  if (s.claude_status === 'working') return 'working';
  return 'idle';
}

/** When the row entered the state its bucket describes, best effort. */
function bucketSince(s: SessionRow, bucket: TriageBucket): number {
  switch (bucket) {
    case 'stuck':
      return s.stuck_since ?? s.last_activity_at;
    case 'lifecycle':
      return s.lost_at ?? s.safe_kill_requested_at ?? s.last_activity_at;
    case 'done_unread':
      return s.last_stop_at ?? s.last_turn_at ?? s.last_activity_at;
    case 'working':
      return s.last_activity_at;
    default:
      return s.idle_since ?? s.last_activity_at;
  }
}

/** Where a row sits in the triage queue. Pure, and `now` is injected, so the
 *  sidebar, the tests and later the digest all agree. */
export function rank(s: SessionRow, opts: AttentionOptions): TriageRank {
  const bucket = classify(s, opts);
  const order = TRIAGE_BUCKETS.indexOf(bucket);
  const ageSecs = Math.min(AGE_CAP_SECS - 1, Math.max(0, opts.now - bucketSince(s, bucket)));
  return { bucket, order, ageSecs, score: (TRIAGE_BUCKETS.length - order) * AGE_CAP_SECS + ageSecs };
}

export function needsYou(s: SessionRow, opts: AttentionOptions): boolean {
  return NEEDS_YOU.has(classify(s, opts));
}

/** How many rows are waiting on the operator right now. Narrower than
 *  `needsYou()` on purpose — see NEEDS_YOU_COUNTED_BUCKETS. */
export function countNeedsYou(rows: readonly SessionRow[], opts: AttentionOptions): number {
  let n = 0;
  for (const s of rows) if (NEEDS_YOU_COUNTED.has(classify(s, opts))) n++;
  return n;
}

/** Rows worst-first: bucket, then the longest wait, then id — so the order is
 *  stable across ticks and never reshuffles under the cursor. */
export function byTriage(rows: readonly SessionRow[], opts: AttentionOptions): SessionRow[] {
  return rows
    .map((s) => ({ s, score: rank(s, opts).score }))
    .sort((a, b) => b.score - a.score || a.s.id - b.s.id)
    .map((x) => x.s);
}

// ── severity (for sorting projects by worst child) ──

/** Higher = worse, derived from the triage buckets so the project tree and the
 *  Needs-you queue can never disagree. Classified with the idle rule off, which
 *  keeps severity a pure function of the row with no clock to inject. */
export function severity(s: SessionRow): number {
  return TRIAGE_BUCKETS.length - TRIAGE_BUCKETS.indexOf(classify(s, { idleSecs: 0, now: 0 }));
}

/** project_id → max severity over its sessions. */
export function worstSeverityByProject(sessions: readonly SessionRow[]): Map<number, number> {
  const out = new Map<number, number>();
  for (const s of sessions) {
    if (s.project_id == null) continue;
    const sev = severity(s);
    if (sev > (out.get(s.project_id) ?? -1)) out.set(s.project_id, sev);
  }
  return out;
}

// ── stuck transitions ──

/** session.id → stuck_kind for every currently-stuck row. */
export function stuckSnapshot(sessions: readonly SessionRow[]): Map<number, StuckKind> {
  const m = new Map<number, StuckKind>();
  for (const s of sessions) if (s.stuck_kind) m.set(s.id, s.stuck_kind);
  return m;
}

/** Rows that became stuck (or changed stuck kind) since `prev`. Clearing is
 *  not a transition worth announcing. */
export function newlyStuck(
  prev: ReadonlyMap<number, StuckKind>,
  sessions: readonly SessionRow[],
): SessionRow[] {
  const out: SessionRow[] = [];
  for (const s of sessions) {
    if (!s.stuck_kind) continue;
    if (prev.get(s.id) !== s.stuck_kind) out.push(s);
  }
  return out;
}

/** The label a row shows in the sidebar / notifications. */
export function displayName(s: SessionRow, friendly: boolean): string {
  return friendly && s.friendly_name ? s.friendly_name : s.tmux_name;
}

/** One-line announcement for a stuck transition. */
export function stuckMessage(s: SessionRow, friendly = true): string {
  return `${displayName(s, friendly)} on ${s.host_alias} is stuck: ${stuckKindLabel(s.stuck_kind)}`;
}

// ── outcome display ──

/** Compact elapsed duration: "42s", "5m", "3h 12m", "2d 4h". */
export function formatElapsed(fromUnix: number | null, nowUnix: number): string {
  if (fromUnix === null) return '—';
  const secs = Math.max(0, nowUnix - fromUnix);
  if (secs < 60) return `${secs}s`;
  const mins = Math.floor(secs / 60);
  if (mins < 60) return `${mins}m`;
  const hours = Math.floor(mins / 60);
  if (hours < 24) return `${hours}h ${mins % 60}m`;
  const days = Math.floor(hours / 24);
  return `${days}d ${hours % 24}h`;
}

/** The instant a session's clock starts: fleet's `started_at`, else tmux's
 *  `created_at`. */
export function sessionStart(s: Pick<SessionRow, 'started_at' | 'created_at'>): number {
  return s.started_at ?? s.created_at;
}

/** First line of the last prompt, ellipsised to `max` chars, for a row's
 *  secondary text. */
export function promptPreview(prompt: string | null, max = 60): string {
  if (!prompt) return '';
  const line = prompt.split('\n').find((l) => l.trim().length > 0)?.trim() ?? '';
  const chars = Array.from(line);
  return chars.length > max ? chars.slice(0, max - 1).join('') + '…' : line;
}

export function ciStatusLabel(status: CiStatus | null): string {
  switch (status) {
    case 'passing':
      return '✓ CI';
    case 'failing':
      return '✗ CI';
    case 'pending':
      return '… CI';
    default:
      return '';
  }
}

export function ciStatusColor(status: CiStatus | null): string {
  switch (status) {
    case 'passing':
      return '#50c86e';
    case 'failing':
      return '#e64a4a';
    case 'pending':
      return '#d29b4a';
    default:
      return 'transparent';
  }
}

File: crates/fleet-core/src/service/attention.rs
//! Which sessions need a person, decided once, here.
//!
//! # Why this is on the hub
//!
//! "What needs me?" is the question a phone is opened to answer, and until
//! now every client worked it out for itself from rows it had to download
//! first. The desktop has the rich version (`src/lib/attention.ts`,
//! `classify`), fleet-mobile has a two-condition predicate, and the hub's own
//! `fleet_health` roll-up has no notion of it at all — it counts ghosts,
//! stuck rows and context pressure, never "a person is needed here".
//!
//! So a phone downloaded 44 full rows — 51 968 B measured — to find the three
//! that wanted an answer. With the reason on the row and a filter beside it,
//! that is 1 668 B.
//!
//! The divergence between those two classifiers is latent rather than
//! absent: replayed over a 56-row capture they agree exactly, but only
//! because that capture holds no `failed` row, no ghost, no `lost_at` and no
//! `safe_kill_state`. The first one of those would have split them.
//!
//! # What is here and what is not
//!
//! Only the classification. Ranking, scoring and sort order stay with the
//! client: they are how a screen chooses to present the queue, not a fact
//! about the fleet.
//!
//! The desktop's "idle for too long" bucket is deliberately **not** decided
//! here either. It depends on an operator-configured idle threshold,
//! and a hub that guessed one would quietly disagree with the desktop that
//! set it. What this module answers is the states that need no knob to
//! recognise; a client is free to add its own idle rule on top.

use crate::store::SessionRow;

/// Why a session needs a person. The order is the urgency order, and it is
/// the same order the desktop's `TRIAGE_BUCKETS` uses, so the two cannot
/// drift into disagreeing about which of two reasons wins.
#[derive(Debug, Clone, Copy, PartialEq, Eq, serde::Serialize, serde::Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum Reason {
    /// Blocked on a dialog: a permission prompt, a question, an elicitation.
    /// The one state where an answer is all that is wanted.
    Waiting,
    /// Wedged in a way the REPL will not leave on its own — an auth menu, a
    /// reconnect, an OOM. `stuck_kind` says which.
    Stuck,
    /// Claude reported a failed turn.
    Failed,
    /// The session's lifecycle is broken: a safe kill that failed or is still
    /// pending, a ghost row, or a row the fleet has lost track of.
    Lifecycle,
}

impl Reason {
    /// The wire spelling, matching the desktop's bucket names.
    pub fn as_str(self) -> &'static str {
        match self {
            Reason::Waiting => "waiting",
            Reason::Stuck => "stuck",
            Reason::Failed => "failed",
            Reason::Lifecycle => "lifecycle",
        }
    }
}

/// A session that needs a person, and since when.
#[derive(Debug, Clone, Copy, PartialEq, Eq, serde::Serialize, serde::Deserialize)]
pub struct Attention {
    pub reason: Reason,
    /// Best-effort unix second the session entered this state. Falls back to
    /// `last_activity_at`, which is always present, so a client can always
    /// draw an age.
    pub since: i64,
}

/// Whether this row needs a person, and why.
///
/// The order of the checks *is* the precedence: a session that is both
/// blocked and ghosted is reported as blocked, because that is the one a
/// person can do something about right now.
///
/// An `external` session — a Claude running outside fleet entirely — never
/// qualifies, whatever its fields say: it is read-only here, so reporting it
/// as needing a person offers an action that does not exist.
pub fn needs_attention(row: &SessionRow) -> Option<Attention> {
    if row.kind == "external" {
        return None;
    }
    let reason = if row.claude_status.as_deref() == Some("blocked") {
        Reason::Waiting
    } else if row.stuck_kind.is_some() {
        Reason::Stuck
    } else if row.claude_status.as_deref() == Some("failed") {
        Reason::Failed
    } else if is_lifecycle_broken(row) {
        Reason::Lifecycle
    } else {
        return None;
    };
    Some(Attention {
        reason,
        since: since_for(row, reason),
    })
}

fn is_lifecycle_broken(row: &SessionRow) -> bool {
    matches!(
        row.safe_kill_state.as_deref(),
        Some("failed") | Some("requested")
    ) || row.status == "ghost"
        || row.lost_at.is_some()
}

fn since_for(row: &SessionRow, reason: Reason) -> i64 {
    match reason {
        Reason::Stuck => row.stuck_since.unwrap_or(row.last_activity_at),
        Reason::Lifecycle => row
            .lost_at
            .or(row.safe_kill_requested_at)
            .unwrap_or(row.last_activity_at),
        _ => row.last_activity_at,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    /// A plain running session that needs nobody, to vary one field at a time.
    fn row() -> SessionRow {
        SessionRow {
            id: 1,
            row_version: 0,
            tmux_name: "t".to_string(),
            host_alias: "alpha".to_string(),
            project_id: None,
            worktree_id: None,
            created_at: 0,
            last_activity_at: 100,
            status: "running".to_string(),
            notes: None,
            account_uuid: None,
            kind: "work".to_string(),
            reviews_session_id: None,
            worktree_key: None,
            lost_at: None,
            lost_reason: None,
            claude_session_id: None,
            claude_status: Some("working".to_string()),
            effort_level: None,
            pr_url: None,
            current_activity: None,
            context_pct: None,
            stuck_kind: None,
            friendly_name: None,
            safe_kill_state: None,
            safe_kill_nonce: None,
            safe_kill_detail: None,
            safe_kill_requested_at: None,
            idle_since: None,
            stuck_since: None,
            last_playbook_at: None,
            last_prompt: None,
            started_at: None,
            last_turn_at: None,
            ci_status: None,
            turn_seq: 0,
            last_stop_at: None,
            parent_session_id: None,
            tags: Vec::new(),
            usage: Default::default(),
            context: Default::default(),
            pending_input: None,
            work: None,
            work_rejected: vec![],
        }
    }

    #[test]
    fn a_blocked_session_is_waiting_and_an_ordinary_one_is_nothing() {
        let mut r = row();
        assert_eq!(needs_attention(&r), None, "a working session needs nobody");

        r.claude_status = Some("blocked".into());
        assert_eq!(needs_attention(&r).unwrap().reason, Reason::Waiting);
    }

    /// The order of the checks is the precedence, and this is the pair that
    /// makes it matter: a person can answer a blocked session now, and can do
    /// nothing about a ghost row until they are at a terminal.
    #[test]
    fn blocked_outranks_every_other_reason() {
        let mut r = row();
        r.claude_status = Some("blocked".into());
        r.stuck_kind = Some("auth_menu".into());
        r.status = "ghost".into();
        assert_eq!(needs_attention(&r).unwrap().reason, Reason::Waiting);
    }

    #[test]
    fn stuck_failed_and_the_three_lifecycle_shapes_each_qualify() {
        let mut r = row();
        r.stuck_kind = Some("oom".into());
        assert_eq!(needs_attention(&r).unwrap().reason, Reason::Stuck);

        let mut r = row();
        r.claude_status = Some("failed".into());
        assert_eq!(needs_attention(&r).unwrap().reason, Reason::Failed);

        for broken in [
            |r: &mut SessionRow| r.safe_kill_state = Some("failed".into()),
            |r: &mut SessionRow| r.safe_kill_state = Some("requested".into()),
            |r: &mut SessionRow| r.status = "ghost".into(),
            |r: &mut SessionRow| r.lost_at = Some(99),
        ] {
            let mut r = row();
            broken(&mut r);
            assert_eq!(
                needs_attention(&r).map(|a| a.reason),
                Some(Reason::Lifecycle)
            );
        }
    }

    /// An external session is a Claude running outside fleet: nothing here can
    /// act on it, so reporting it would offer an action that does not exist.
    #[test]
    fn an_external_session_never_needs_a_person() {
        let mut r = row();
        r.kind = "external".into();
        r.claude_status = Some("blocked".into());
        r.stuck_kind = Some("oom".into());
        assert_eq!(needs_attention(&r), None);
    }

    /// A client draws an age from `since`, so it must always be a number.
    #[test]
    fn since_prefers_the_state_stamp_and_always_falls_back() {
        let mut r = row();
        r.last_activity_at = 100;
        r.stuck_kind = Some("oom".into());
        assert_eq!(
            needs_attention(&r).unwrap().since,
            100,
            "no stuck_since yet"
        );

        r.stuck_since = Some(140);
        assert_eq!(needs_attention(&r).unwrap().since, 140);

        let mut r = row();
        r.last_activity_at = 100;
        r.lost_at = Some(150);
        assert_eq!(needs_attention(&r).unwrap().since, 150);
    }

    /// The wire spellings are the desktop's bucket names; a rename on one
    /// side without the other would split the two classifiers silently.
    #[test]
    fn the_wire_spellings_match_the_desktop_buckets() {
        let ts = include_str!("../../../../src/lib/attention.ts");
        for r in [
            Reason::Waiting,
            Reason::Stuck,
            Reason::Failed,
            Reason::Lifecycle,
        ] {
            assert!(
                ts.contains(&format!("'{}'", r.as_str())),
                "src/lib/attention.ts does not name the bucket {}",
                r.as_str()
            );
        }
    }
}
